File: SampleSolution.java
import java.util.*;

public class SampleSolution {
    private int totalTilt = 0;

    public int findTilt(TreeNode root) {
        this.totalTilt = 0;
        this.findTiltHelper(root);
        return this.totalTilt;
    }

    private int findTiltHelper(TreeNode node) {
        if (node == null)
            return 0;

        int leftSum = this.findTiltHelper(node.left);
        int rightSum = this.findTiltHelper(node.right);
        int tilt = Math.abs(leftSum - rightSum);
        this.totalTilt += tilt;

        return node.val + leftSum + rightSum;
    }

    public class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;
        TreeNode() {}
        TreeNode(int val) { this.val = val; }
        TreeNode(int val, TreeNode left, TreeNode right) {
            this.val = val;
            this.left = left;
            this.right = right;
        }
    }

    public static String freqAlphabets(String s) {
        char[] ch = s.toCharArray();
        char[] c = new char[s.length()];
        int a = 0;
        for (int i = 0; i < s.length(); i++) {
            if (i + 2 < s.length() && ch[i + 2] == '#') {
                c[a++] = (char) ((ch[i] - '0') * 10 + (ch[i + 1] - '1') + 'a');
                i += 2;
            } else {
                c[a++] = (char) ((ch[i] - '1') + 'a');
            }
        }
        String st = String.valueOf(c).trim();
        return st;
    }

    public static List<String> buildArray(int[] target, int n) {
        List<String> res = new ArrayList<>();
        int index = 0;
        for (int i = 0; i < n && index < target.length; i++) {
            if (target[index] == i + 1) {
                res.add("Push");
                index++;
            } else {
                res.add("Push");
                res.add("Pop");
            }
        }
        return res;
    }

    public static int findJudge(int n, int[][] trust) {
        if (n == 1)
            return 1;
        int[] truster = new int[n];
        int[] trustee = new int[n];
        for (int i = 0; i < trust.length; i++) {
            truster[trust[i][0] - 1]++;
            trustee[trust[i][1] - 1]++;
        }

        for (int i = 0; i < n; i++) {
            if (trustee[i] == n - 1 && truster[i] == 0) {
                return i + 1;
            }
        }

        return -1;
    }

    public static String[] reorderLogFiles(String[] logs) {
        Comparator<String> myComp = new Comparator<String>() {
            @Override
            public int compare(String s1, String s2) {
                int s1SpaceIndex = s1.indexOf(' ');
                int s2SpaceIndex = s2.indexOf(' ');
                char s1FirstCharacter = s1.charAt(s1SpaceIndex + 1);
                char s2FirstCharacter = s2.charAt(s2SpaceIndex + 1);

                if (s1FirstCharacter <= '9') {
                    if (s2FirstCharacter <= '9')
                        return 0;
                    else
                        return 1;
                }
                if (s2FirstCharacter <= '9')
                    return -1;

                int preCompute = s1.substring(s1SpaceIndex + 1).compareTo(s2.substring(s2SpaceIndex + 1));
                if (preCompute == 0)
                    return s1.substring(0, s1SpaceIndex).compareTo(s2.substring(0, s2SpaceIndex));
                return preCompute;
            }
        };
        Arrays.sort(logs, myComp);
        return logs;
    }

    public List<List<Integer>> combinationSum2(int[] candidates, int target) {
        List<List<Integer>> res = new ArrayList<>();
        if (candidates.length == 0)
            return res;

        int[] counts = new int[51];
        for (int c : candidates)
            counts[c]++;

        combinationSum2Helper(counts, res, target, new ArrayList<Integer>(), 1);

        return res;
    }

    private void combinationSum2Helper(int[] counts, List<List<Integer>> res, int target, List<Integer> path,
            int index) {
        if (target == 0) {
            res.add(new ArrayList<>(path));
            return;
        }

        for (int i = index; i < counts.length; i++) {
            if (i <= target && counts[i] > 0) {
                counts[i]--;
                path.add(i);
                combinationSum2Helper(counts, res, target - i, path, i);
                path.remove(path.size() - 1);
                counts[i]++;
            } else if (i > target)
                return;
        }
    }

    public static int minMaxGame(int[] nums) {
        if (nums.length == 1)
            return nums[0];
        int[] newNums = new int[nums.length / 2];
        for (int i = 0; i < nums.length / 2; i++) {
            if (i % 2 == 0)
                newNums[i] = Math.min(nums[2 * i], nums[2 * i + 1]);
            else
                newNums[i] = Math.max(nums[2 * i], nums[2 * i + 1]);
        }
        return minMaxGame(newNums);
    }

    public static int findLHS(int[] nums) {
        HashMap<Integer, Integer> map = new HashMap<>();
        int res = 0;
        for (int num : nums) {
            map.put(num, map.getOrDefault(num, 0) + 1);
        }
        for (int key : map.keySet()) {
            if (map.containsKey(key + 1))
                res = Math.max(res, map.get(key) + map.get(key + 1));
        }
        return res;
    }

    public static int romanToInt(String S) {
        int ans = 0, num = 0;

        for (int i = S.length() - 1; i >= 0; i--) {
            switch (S.charAt(i)) {
                case 'I':
                    num = 1;
                    break;
                case 'V':
                    num = 5;
                    break;
                case 'X':
                    num = 10;
                    break;
                case 'L':
                    num = 50;
                    break;
                case 'C':
                    num = 100;
                    break;
                case 'D':
                    num = 500;
                    break;
                case 'M':
                    num = 1000;
                    break;
            }
            if (4 * num < ans)
                ans -= num;
            else
                ans += num;
        }
        return ans;
    }

    public static int countTriples(int n) {
        int x = 2;
        int sum = 0;

        while (true) {
            for (int y = 1; y < x; y++) {
                if (gcd(x, y) != 1 || ((x % 2) == 1 && (y % 2) == 1)) {
                    continue;
                }

                int res = x * x + y * y;
                if (res <= n) {
                    sum += n / res;
                } else if (y == 1) {
                    return 2 * sum;
                }
            }
            x++;
        }
    }

    public static int gcd(int a, int b) {
        while (b != 0) {
            int r = a % b;
            a = b;
            b = r;
        }
        return a;
    }

    public static boolean isAnagram(String s, String t) {
        int[] freq = new int[26];

        for (char c : s.toCharArray()) {
            freq[c - 'a'] += 1;
        }

        for (char c : t.toCharArray()) {
            freq[c - 'a'] -= 1;
        }

        for (int i : freq) {
            if (i != 0) {
                return false;
            }
        }

        return true;
    }
}

File: SampleSolution.py
from collections import Counter
import math
from typing import List


class SampleSolution:
    def canMakeArithmeticProgression(self, arr: List[int]) -> bool:
        arr.sort()
        diff=abs(arr[0] - arr[1])
        for i in range(1,len(arr)-1):
            if abs(arr[i] - arr[i+1]) != diff:
                return False
        return True
    
    def numDifferentIntegers(self, word: str) -> int:
        word = word + 'a'
        ht = dict() 
        
        tmpStr = ""
        for ch in word:
            if not 'a' <= ch <= 'z':
                tmpStr += ch
            else:
                if tmpStr != "":
                    ht[int(tmpStr)] = 1
                    tmpStr = "" 
        return len(ht)
    
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        hashmap = {}
        for i in range(len(nums)):
            complement = target - nums[i]
            if complement in hashmap:
                return [i, hashmap[complement]]
            hashmap[nums[i]] = i
    
    def arrayStringsAreEqual(self, word1: List[str], word2: List[str]) -> bool:
        return "".join(word1) == "".join(word2)
    
    def checkZeroOnes(self, s: str) -> bool:
        cnt_one, max_one = 0, 0
        cnt_zero, max_zero = 0, 0
        prev = None
        
        for c in s:
            if c == "1":
                cnt_one += 1
                max_one = max(max_one, cnt_one)
                cnt_zero = 0

            else:
                cnt_zero += 1
                max_zero = max(max_zero, cnt_zero)
                cnt_one = 0
            
            prev = c
        
        return max_one > max_zero
    
    def maxProfit(self, prices: List[int]) -> int:
        minPrice, maxProfit = float(math.inf), 0
        for price in prices:
            if price < minPrice:
                minPrice = price

            curProfit = price - minPrice
            if curProfit > maxProfit:
                maxProfit = curProfit

        return maxProfit
    class TreeNode:
        def __init__(self, val=0, left=None, right=None):
            self.val = val
            self.left = left
            self.right = right
    
    def findTilt(self, root: TreeNode) -> int:
        total_tilt = 0

        def valueSum(node):
            nonlocal total_tilt

            if not node:
                return 0

            left_sum = valueSum(node.left)
            right_sum = valueSum(node.right)
            tilt = abs(left_sum - right_sum)
            total_tilt += tilt

            return left_sum + right_sum + node.val

        valueSum(root)

        return total_tilt
    
    def freqAlphabets(self, s: str) -> str:
        i = len(s) - 1
        ans = []
        while i >= 0:
            if s[i] == '#':
                ans.append(chr(ord('`')+int(s[i-2:i])))
                i -= 3
            else:
                ans.append(chr(ord('`')+int(s[i])))
                i -= 1
        return ''.join(reversed(ans))
    
    def buildArray(self, target: List[int], n: int) -> List[str]:
        target_set = set(target)
        maximum = min(target[-1], n)
        output = []
        for i in range(1, maximum + 1):
            if i in target_set:
                output.append("Push")
            else:
                output.extend(["Push", "Pop"])
        
        return output
    
    def findJudge(self, n: int, trust: List[List[int]]) -> int:
        if not trust and n==1: return 1
        if len(trust) < n-1: return -1
        l1 = [a for a, b in trust]
        l2 = set(b for a,b in trust)
        miss_vals = set(range(1,n+1)) - set(l1)
        for val in miss_vals:
            l3 = [a for a, b in trust if b == val]
            if not l3: continue
            l3.append(val)
            l3.sort()
            if l3 == list(range(1,n+1)):
                return val            
        return -1
    
    def reorderLogFiles(self, logs: List[str]) -> List[str]:
        digits = []
        letters = []
        
        for l in logs:
            if l.split(" ")[1].isdigit():
                digits.append(l)
            else:
                letters.append(l)
        
        letters.sort(key=lambda log: log.split(" ", 1)[0])
        letters.sort(key=lambda log: log.split(" ", 1)[1])
        return letters + digits
    
    def combinationSum2(self, candidates: List[int], target: int) -> list[List[int]]:
        def backtrack(comb, remain, curr, results):

            if remain == 0:
                # make a deep copy of the resulted combination
                results.append(list(comb))
                return

            for next_curr in range(curr, len(candidates)):

                if next_curr > curr \
                  and candidates[next_curr] == candidates[next_curr-1]:
                    continue

                pick = candidates[next_curr]
                # optimization: skip the rest of elements starting from 'curr' index
                if remain - pick < 0:
                    break

                comb.append(pick)
                backtrack(comb, remain - pick, next_curr + 1, results)
                comb.pop()

        candidates.sort()

        comb, results = [], []
        backtrack(comb, target, 0, results)

        return results
    
    def minMaxGame(self, nums: List[int]) -> int:
        while len(nums) > 1:
            newnums = []
            for i in range(len(nums)//2):
                if i%2 == 0:
                    newnums.append(min(nums[2*i],nums[2*i+1]))
                else:
                    newnums.append(max(nums[2*i],nums[2*i+1]))
            nums = newnums
            
        return nums[0]
    
    def findLHS(self, nums: list[int]) -> int:
        c = Counter(nums)
        largestlength = 0 
        for n in c: 
            if n+1 in c: 
                largestlength = max(largestlength, c[n]+c[n+1])
                
        return largestlength
    
    def romanToInt(self, s: str) -> int:
        roman_table = {"I":1,"V":5,"X":10,"L":50,"C":100,"D":500,"M":1000}
        num = 0
        last = "I"
        for numeral in s[::-1]:
            if roman_table[numeral]<roman_table[last]:
                num -= roman_table[numeral]
            else:
                num += roman_table[numeral]
            last = numeral
        return num
    
    def countTriples(self, x: int) -> int:
        result = 0
        xsqrt = math.ceil(math.sqrt(x))
        gcd = math.gcd
        for m in range(1, xsqrt, 2):
            for n in range(2, xsqrt, 2):
                if gcd(m, n) == 1:
                    c = m ** 2 + n ** 2
                    result = result + x // c
        return 2 * result
    
    def isAnagram(self, s: str, t: str) -> bool:
        for c in 'abcdefghijklmnopqrstuvwxyz':
            if s.count(c) != t.count(c): return False
        return True